explorer/dir_android.go

```diff
@@ -0,0 +1,122 @@
+package explorer
+
+import (
+	"gioui.org/app"
+	"git.wow.st/gmp/jni"
+	"os"
+)
+
+//go:generate javac -source 8 -target 8  -bootclasspath $ANDROID_HOME/platforms/android-30/android.jar -d $TEMP/explorer_file_android/classes dir_android.java
+//go:generate jar cf dir_android.jar -C $TEMP/explorer_file_android/classes .
+
+type Directory struct {
+	createFile jni.MethodID
+	libClass   jni.Class
+	libObject  jni.Object
+
+	url      jni.Object
+	openFile jni.MethodID
+}
+
+func newDirectory(env jni.Env, url jni.Object) (*Directory, error) {
+	f := &Directory{url: url}
+
+	class, err := jni.LoadClass(env, jni.ClassLoaderFor(env, jni.Object(app.AppContext())), "org/gioui/x/explorer/dir_android")
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := jni.NewObject(env, class, jni.GetMethodID(env, class, "<init>", `()V`))
+	if err != nil {
+		return nil, err
+	}
+
+	// For some reason, using `f.stream` as argument for a constructor (`public file_android(Object j) {}`) doesn't work.
+	if err := jni.CallVoidMethod(env, obj, jni.GetMethodID(env, class, "setHandle", `(Ljava/lang/Object;)V`), jni.Value(f.url)); err != nil {
+		return nil, err
+	}
+
+	f.libObject = jni.NewGlobalRef(env, obj)
+	f.libClass = jni.Class(jni.NewGlobalRef(env, jni.Object(class)))
```

Comment:
Where and how are these global references cleaned up? I don't see any mechanism to do that, and I don't think the JVM is able to do it automatically.